OurUmbraco/Auth/AppBuilderExtensions.cs

```diff
@@ -0,0 +1,130 @@
+                        if (!TryGetHeaderValue(context, ProjectAuthConstants.ProjectIdHeader, out var projectId))
+                            throw new InvalidOperationException("No project Id found in request"); // this will never happen
+                        if (!TryGetHeaderValue(context, ProjectAuthConstants.MemberIdHeader, out var memberId))
+                            throw new InvalidOperationException("No project Id found in request"); // this will never happen
+
+                        //Get the stored auth key for this project and member
+                        var tokenService = new ProjectAuthKeyService(ApplicationContext.Current.DatabaseContext);
+                        var authKeys = tokenService.GetAllAuthKeysForProject(projectId);
+
+                        var timestamp = new DateTime();
+
+                        if (authKeys.Any(authKey =>
+                            HMACAuthentication.ValidateToken(context.Token, requestPath, authKey, out timestamp)))
+                        {
+                            //If ok, create a ticket here with the Claims we need to check for in AuthZ
+                            var ticket = new AuthenticationTicket(
+                                new ClaimsIdentity(
+                                    new List<Claim>
+                                    {
+                                        new Claim(ProjectAuthConstants.BearerTokenClaimType, ProjectAuthConstants.BearerTokenClaimValue),
+                                        new Claim(ProjectAuthConstants.ProjectIdClaim, projectId.ToInvariantString()),
+                                        new Claim(ProjectAuthConstants.MemberIdClaim, memberId.ToInvariantString()),
+                                    },
+
+                                    //The authentication type = this is important, if not set
+                                    //then the ticket's IsAuthenticated property will be false
+                                    authenticationType: ProjectAuthConstants.BearerTokenAuthenticationType),
+                                new AuthenticationProperties
+                                {
+                                    //Expires after 5 minutes in case there are some long running operations
+                                    ExpiresUtc = timestamp.AddMinutes(5)
+                                });
+
+                            context.SetTicket(ticket);
+                        }
+                        else
+                        {
+                            throw new InvalidOperationException("Token validation failed");
+                        }
+                    }
+                }
+            });
+        }
+
+        private static bool EnsureHeaderValues(BaseContext context, params string[] headers)
+        {
+            foreach(var h in headers)
+            {
+                if (!TryGetHeaderValue(context, h, out _))
```

Comment:
I never understood this notation, can we add a code comment or rewrite it as something a bit more easy to read?

Reply:
Added a note here: https://github.com/umbraco/OurUmbraco/pull/552/commits/2abb6b9c2ba681353753bbf957f05fd6041d6afb